pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,185 @@
+	podInformer           cache.SharedIndexInformer
+	networkPolicyInformer cache.SharedIndexInformer
+
+	eventChan chan Event
+}
+
+const AddEvent = "Added"
+const DeleteEvent = "Deleted"
+const UpdateEvent = "Updated"
+
+const NetworkPolicy = "np"
+const Pod = "pod"
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    string
+	ObjType string
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) SetEventChannel(eventChan chan Event) {
+	rw.eventChan = eventChan
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rw.stopCh, rw.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rw.stopCh, rw.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rw)
+		}
+	}()
+}
+
+func (rw *RemoteCluster) GetPods() []interface{} {
+	return rw.podInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) GetNetworkPolicies() []interface{} {
+	return rw.networkPolicyInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) OnAdd(obj interface{}) {
+	objType := ""
+
+	if pod, ok := obj.(*v1.Pod); ok {
+		klog.Infof("Pod discovered in %s: %s", rw.ClusterID, pod.GetSelfLink())
+		objType = Pod
+	} else if np, ok := obj.(*v1net.NetworkPolicy); ok {
+		klog.Infof("NetworkPolicy discovered in %s: %s", rw.ClusterID, np.GetSelfLink())
+		objType = NetworkPolicy
+	} else {
+		klog.Errorf("OnAdd for unexpected object type: %v", obj)
+		return
+	}
+
+	if rw.eventChan != nil {
+		rw.eventChan <- Event{
+			Cluster: rw,
+			Type:    AddEvent,
+			ObjType: objType,
+			Objs:    []interface{}{obj},
+		}
+	}
```

Comment:
Why?, isn't it a pointer?, once it's not nil I think we can assume it's complete.

Reply:
`eventChan` can be accessed by multiple threads so there must be synchronization between them to guarantee that a write done by one thread is visible to other threads. Otherwise, if thread 1 executes `SetEventChannel`, it's possible that thread 2 (or thread 3 etc) executing `OnAdd` never sees the write of `eventChan`, due to CPU caching.  So you need some form of synchronization or a memory barrier like a Mutex to guarantee visibility of the write.

Reply:
Ah, ok I understand what you mean. I had a lock for this at some point, but as I changed structure, that's gone. Let me add a similar mechanism which could be controlled from outside (the controller/consumer).

Reply:
Done.

Reply:
> oh, ok this is what you mean: https://notes.shichao.io/gopl/ch9/#memory-synchronization
> 
> So mutex operations make the current core flush cache, I'll do that.

Essentially yes or not cache it and always retrieve from main memory. The `volatile` keyword in Java instructs the compiler to ensure proper memory synchronization as well but Go doesn't have that, although it does provide low-level [atomic](https://golang.org/pkg/sync/atomic/) synchronization primitives. Of course Go has been designed to discourage the use of shared memory if at all possible and use channels instead for concurrency so you don't have too worry about memory race conditions.